### PR TITLE
Add support for REST Collections API to bulk data tasks

### PR DESCRIPTION
**Pull request by davidmreed**
# Critical Changes

# Changes

- CumulusCI now supports using the REST Collections API in data load, extract, and delete operations. By default, CumulusCI will select an API for you based on data volume (<2000 records uses the REST API, >=2000 uses Bulk); a desired API can be configured via the mapping file.

# Issues Closed

### CODE REVIEW

File: cumulusci/tasks/bulkdata/load.py [hunk 1]
```diff
@@ -142,42 +144,58 @@ def _execute_step(
...
+                sobject=mapping["sf_object"],
+                operation=(
+                    DataOperationType.INSERT
+                    if mapping.get("action") == "insert"
+                    else DataOperationType.UPDATE
+                ),
+                api_options={"batch_size": mapping.batch_size},
+                context=self,
+                fields=self._get_columns(mapping),
+            )
```
Comment (davisagli) [inline cumulusci/tasks/bulkdata/load.py:180]: I would guess this logic for picking the API/operation will be easier to test if you refactor it into a separate method.

File: cumulusci/tasks/bulkdata/step.py [hunk 2]
```diff
@@ -220,6 +231,39 @@ def get_results(self):
...
+    def query(self):
+        pass
+
+    def get_results(self):
+        for chunk in get_batch_iterator(iter(self.ids), 2000):
+            yield from self.sf.restful(
+                f"composite/sobjects/{self.sobject}",
+                method="POST",
+                json={"ids": chunk, "fields": self.fields},
+            )
```
Comment (davisagli) [inline cumulusci/tasks/bulkdata/step.py:264]: It doesn't look like RestApiRetrieveOperation is used anywhere?

File: cumulusci/tasks/bulkdata/load.py [hunk 3]
```diff
@@ -142,42 +144,58 @@ def _execute_step(
...
 
         statics = self._get_statics(mapping)
-        query = self._query_db(mapping)
 
         total_rows = 0
 
-        # 10,000 is the maximum Bulk API size. Clamping the yield from the query ensures we do not
-        # create more Bulk API batches than expected, regardless of batch size, while capping
-        # memory usage.
-        for row in query.yield_per(10000):
```
Comment (davisagli) [inline cumulusci/tasks/bulkdata/load.py:180]: I know the bulk API operations now have improved logic for creating batches that fit within limits, but this is still useful for making sure we don't try to load all rows from the SQL db into memory at once, which can be a problem when processing millions of records.

File: cumulusci/tasks/bulkdata/step.py (hunk 2 above)
Comment (davidmreed) [reply cumulusci/tasks/bulkdata/step.py:264]: I thought I needed it to support Scott's Person Account use case but was mistaken I'll remove it if I don't find a good use for it - it's so nice to be able to pull 2k records at a transaction.

File: cumulusci/tasks/bulkdata/load.py (hunk 3 above)
Comment (davidmreed) [reply cumulusci/tasks/bulkdata/load.py:176]: I've restored this element.

File: cumulusci/tasks/bulkdata/delete.py [hunk 4]
```diff
@@ -55,6 +56,19 @@ def _init_options(self, kwargs):
...
+        try:
+            self.options["api"] = {
+                "bulk": DataApi.BULK,
+                "rest": DataApi.REST,
+                "smart": DataApi.SMART,
+            }[self.options.get("api", "smart").lower()]
+        except KeyError:
+            raise TaskOptionsError(
+                f"{self.options['api']} is not a valid value for API (valid: bulk, rest, smart"
+            )
```
Comment (davisagli) [inline cumulusci/tasks/bulkdata/delete.py:68]: `api` should be included in `task_options` above (and task docs regenerated)

File: cumulusci/tasks/bulkdata/load.py [hunk 5]
```diff
@@ -138,41 +139,43 @@ def _execute_step(
 
         mapping["oid_as_pk"] = bool(mapping.get("fields", {}).get("Id"))
 
+        local_ids = []
+        query = self._query_db(mapping)
+        operation = (
+            DataOperationType.INSERT
+            if mapping.get("action") == "insert"
+            else DataOperationType.UPDATE
+        )
```
Comment (davisagli) [inline cumulusci/tasks/bulkdata/load.py:148]: Perhaps the mapping parser should be updated to parse `action` into one of these two types.

File: cumulusci/tasks/bulkdata/step.py [hunk 6]
```diff
@@ -220,6 +232,32 @@ def get_results(self):
...
+        super().__init__(
+            sobject=sobject, api_options=api_options, context=context, query=query
+        )
+        self.fields = fields
+
+    def query(self):
+        # This is not as memory-efficient as we might like.
+        # To preserve the invariant from the Bulk API that
+        # we know the query count before consuming the results,
+        # we use `query_all` instead of `query_all_iter`.
```
Comment (davisagli) [inline cumulusci/tasks/bulkdata/step.py:248]: Couldn't you get the first batch with `query` and grab the `totalSize`, and then make get_results call `query_more` repeatedly the same way that `query_all_iter` does?

File: cumulusci/tasks/bulkdata/step.py [hunk 7]
```diff
@@ -365,3 +401,153 @@ def get_results(self):
...
+                r[boolean_field] = bool(r[boolean_field])
+
+            # Remove empty fields (different semantics in REST API)
+            # We do this for insert only - on update, any fields set to `null`
+            # are meant to be blanked out.
+            if self.operation is DataOperationType.INSERT:
+                r = {k: r[k] for k in r if r[k] is not None and r[k] != ""}
+
+            r["attributes"] = {"type": self.sobject}
+            return r
```
Comment (davisagli) [inline cumulusci/tasks/bulkdata/step.py:436]: Consider using longer variable names to make this easier to read.

File: cumulusci/tasks/bulkdata/mapping_parser.py [hunk 8]
```diff
@@ -72,6 +72,8 @@ class MappingStep(CCIDictModel):
     static: Dict[str, str] = {}
     filters: List[str] = []
     action: str = "insert"
+    api: DataApi = DataApi.SMART
+    batch_size: int = 200
     oid_as_pk: bool = False  # this one should be discussed and probably deprecated
     record_type: Optional[str] = None  # should be discussed and probably deprecated
     bulk_mode: Optional[
```
Comment (davisagli) [inline cumulusci/tasks/bulkdata/mapping_parser.py:79]: Should we add a validator to make sure the batch size can't be set too high (i.e. with different limits based on which api is specified)?

File: cumulusci/tasks/bulkdata/step.py [hunk 9]
```diff
@@ -365,3 +401,153 @@ def get_results(self):
...
+
+def get_query_operation(
+    *,
+    sobject: str,
+    fields: List[str],
+    api_options: Dict,
+    context: Any,
+    query: str,
+    api: DataApi,
+) -> BaseQueryOperation:
```
Comment (davisagli) [inline cumulusci/tasks/bulkdata/step.py:497]: docstring

File: cumulusci/tasks/bulkdata/step.py [hunk 10]
```diff
@@ -365,3 +401,153 @@ def get_results(self):
                 raise BulkDataException(
                     f"Failed to download results for batch {batch_id} ({str(e)})"
                 )
+
+
+class RestApiDmlOperation(BaseDmlOperation):
```
Comment (davisagli) [inline cumulusci/tasks/bulkdata/step.py:406]: docstring

File: cumulusci/tasks/bulkdata/step.py [hunk 11]
```diff
@@ -365,3 +401,153 @@ def get_results(self):
...
+def get_dml_operation(
+    *,
+    sobject: str,
+    operation: DataOperationType,
+    fields: List[str],
+    api_options: Dict,
+    context: Any,
+    api: DataApi,
+    volume: int,
+) -> BaseQueryOperation:
```
Comment (davisagli) [inline cumulusci/tasks/bulkdata/step.py:534]: docstring

File: cumulusci/tasks/bulkdata/mapping_parser.py (hunk 8 above)
Comment (davidmreed) [reply cumulusci/tasks/bulkdata/mapping_parser.py:79]: Great idea. Adding.

File: cumulusci/tasks/bulkdata/load.py (hunk 5 above)
Comment (davidmreed) [reply cumulusci/tasks/bulkdata/load.py:148]: Yeah, this one is also held back by the legacy code that constructs mapping steps as dicts (to support `after` steps). This ugliness can be refactored out when I fix that.

File: cumulusci/tasks/bulkdata/step.py (hunk 6 above)
Comment (davidmreed) [reply cumulusci/tasks/bulkdata/step.py]: I was shying away from adding more complexity, but turns out to be quite straightforward - thanks for the hint.

File: cumulusci/tasks/bulkdata/step.py [hunk 12]
```diff
@@ -255,7 +251,14 @@ def get_results(self):
...
 
-        yield from (convert(rec) for rec in self.response["records"])
+        while True:
+            yield from (convert(rec) for rec in self.response["records"])
+            if not self.response["done"]:
+                self.response = self.sf.query_more(
+                    self.response["nextRecordsUrl"], identifier_is_url=True
+                )
+            else:
+                return
```
Comment (davisagli) [inline cumulusci/tasks/bulkdata/step.py:261]:
```suggestion
        while not self.response["done"]:
            yield from (convert(rec) for rec in self.response["records"])
            self.response = self.sf.query_more(
                self.response["nextRecordsUrl"], identifier_is_url=True
            )
```

Comment (davidmreed) [reply cumulusci/tasks/bulkdata/step.py:261]: This would exit early if the query returns a number of results 0<n<batch size, wouldn't it?

Comment (davisagli) [reply cumulusci/tasks/bulkdata/step.py:261]: Yep. I thought it felt too good to be true.

**Review by davisagli — APPROVED** [review]

**Comment by prescod** [comment]
@davidmreed  Just some minor bike shedding: did you consider 'rest', 'bulk' or 'auto'?

**Comment by davidmreed** [comment]
@prescod I didn't think about it; I just liked the tone of 'smart', and I think we want to add other features that could be described that way in the future. Willing to reconsider.

**Comment by prescod** [comment]
Auto rings a bit more "standard" to me, but I don't feel strongly about it. Canadian modesty, I guess.

**Pull request merged**